other/web-js/client/src/client-api.js

```diff
@@ -0,0 +1,212 @@
+import jwtDecode from "jwt-decode";
+
+/**
+ * Connects to a MongoDB Atlas App Services App's client API over HTTPS
+ */
+export class ClientApi {
+  static constructBaseUrl(deployment) {
+    const { deployment_model, cloud, region } = deployment;
+    if (deployment_model === "GLOBAL") {
+      return `https://realm.mongodb.com/api/client/v2.0/`;
+    }
+    if (!cloud || !region) {
+      throw new Error(
+        `Must specify a cloud provider and region for LOCAL Apps. e.g. { "cloud": "aws", "region": "us-east-1" }`
+      );
+    }
+    return `https://${region}.${cloud}.realm.mongodb.com/api/client/v2.0/`;
+  }
+
+  constructor({ appId, deployment_model, cloud, region, onAuthChange }) {
+    this.appId = appId;
+    this.credentialStorage = new CredentialStorage(appId);
+    this.currentUser = this.credentialStorage.get("currentUser");
+    this.deployment = {
+      deployment_model:
+        deployment_model ?? (cloud || region) ? "LOCAL" : "GLOBAL",
+      cloud,
+      region,
+    };
+    this.baseUrl = ClientApi.constructBaseUrl(this.deployment);
+    this.onAuthChange = onAuthChange;
+  }
+
+  // endpointUrl(path: string): string;
+  endpointUrl = (path) => {
+    if (!path.startsWith("/")) {
+      throw new Error(`Client API path must start with a slash ("/")`);
+    }
+    const url = new URL(
+      `app/${this.appId}` + path,
+      this.baseUrl
+    );
+    return url.href;
+  };
+
+  // setCurrentUser(user: User | null): void;
+  setCurrentUser = (user) => {
+    this.currentUser = user;
+    this.credentialStorage.set("currentUser", this.currentUser);
+    this.onAuthChange?.(this.currentUser);
+  };
+
+  // registerUser(provider: string, credentials: any): Promise<void>
+  registerUser = async (provider, credentials) => {
+    const url = this.endpointUrl(`/auth/providers/${provider}/register`);
+    const resp = await fetch(url, {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify(credentials),
+    });
+    if (resp.status === 201) {
+      // Status 201 means the user was created. There is no response body.
+      return;
+    } else {
+      throw new ClientApiError(await resp.json());
+    }
+  };
+
+  // logIn(provider: string, credentials: any): Promise<void>
+  logIn = async (provider, credentials) => {
```

Comment:
higher level question that i just noticed upon reviewing this method: 

why are we using the HTTP auth here instead of the Web SDK auth client which'd abstract all this stuff away from us having to implement?

Reply:
Frankly a big takeaway for me is that you probably should just use the web SDK. That said this shows how to use just HTTP to work with the Data API from a web app. Pulling in the web SDK only to use the data api instead feels strange.